BGP integration

mkalcok opened the pull request:
This is relatively large piece of work that enables integration between MicroOVN and external BGP daemons. It's split into multiple commits for easier review, please review them individually. Commit messages and accompanied documentation pages try to explain intended usage and implementation details as much as possible.

Note: This is an experimental feature that, among other things, includes OVN patches not yet accepted in the upstream.

fnordahl commented on microovn/frr/bgp/service.go:
```diff
@@ -22,9 +39,35 @@ func EnableService() error {
+	}
+
+	return redirectBgp(ctx, s, extConnections, extraConfig.Vrf)
 }
```
I don't see us creating the initial BGP daemon config here, which I would expect.

I know this part is less clearly defined, but approaching it will give us much information about what we need to consider in the final implementation.

Please provide a oneshot type call populating the bare minimum similar to what the functional test does:
```
router bgp 1 vrf ovnvrf10
 neighbor eth1-bgp interface remote-as internal
 neighbor eth2-bgp interface remote-as internal
exit
```

mkalcok replied:
I would disagree on this point. If we commit to doing this, we will need to think about:
* How to properly interact with FRR from the MicroOVN codebase
* Do we want to save the running config when we add this?
  * If yes, we need to purge the config on `microovn bgp disable` (or somehow selectively only remove the section we added)
  * If not, do we want to re-create this config on service restart/reboot
* Additional config option in for local ASN (possibly ASNs also for `remote-as`)
* Does user want bgp `unnumbered` or regular explicit peers?

I think that either we are going to make the `microovn enable` config options way too complicated, or we are going to enforce some config options on the user that they may not want.

Overall I think that the upside of us getting involved with the BGP daemon configuration is very little, as any user that wants this feature should be capable of configuring BGP daemon on their own, and the downside is significant in terms of complexity on the code.

mkalcok replied:
After our follow-up offline conversation, I added `asn` config option to `microovn enable bgp` command. When supplied, MicroOVN will set up iBGP unnumbered FRR config. As for the questions I laid out in my previous comment:

> How to properly interact with FRR from the MicroOVN codebase

This actually turned out to be quite simple via executing `vtysh` with multiple `-c` options, effectively achieving "script-like" execution inside the FRR shell.

> Do we want to save the running config when we add this?

yes, it's much easier than trying to re-inject correct config on service restart

> If yes, we need to purge the config on microovn bgp disable (or somehow selectively only remove the section we added)

On disable, current "startup" config is backed up and then replaced with default frr config.

> Additional config option in for local ASN (possibly ASNs also for remote-as)

`asn` option was added and all BGP interfaces are lumped into this ASN, there's no option to define per-interface ASN now. As for `remote-as`, we always use `internal`

> Does user want bgp unnumbered or regular explicit peers?

User can omit the `asn` config option, in which case MicroOVN won't be configuring anything in FRR. This allows users to configure FRR manually if the auto-generated config from MicroOVN wouldn't suite their needs.

fnordahl commented on docs/explanation/bgp-redirect.rst:
```diff
@@ -0,0 +1,91 @@
+========================
+OVN integration with BGP
+========================
```
I absolutely love how you have taken the care to properly document this, even though this is for a experimental feature branch. Thank you!

mkalcok replied:
How else would I know what I submitted here 2 weeks ago? :laughing:

fnordahl commented on tests/test_helper/bgp_utils.bash:
```diff
@@ -0,0 +1,78 @@
+        -c \"neighbor $interface interface remote-as internal\""
+}
+
+# microovn_start_bgp_unnumbered CONTAINER INTERFACE ASN VRF
+#
+# configure FRR bundled with MicroOVN in the CONTAINER, to
+# start BGP in the unnumbered mode, listening on the INTERFACE
+# in the VRF with ASN.
+function microovn_start_bgp_unnumbered() {
+    local container=$1; shift
+    local interface=$1; shift
+    local asn=$1; shift
+    local vrf=$1; shift
+
+    lxc_exec "$container" "microovn.vtysh \
+        -c \"configure\" \
+        -c \"router bgp $asn vrf $vrf\" \
+        -c \"neighbor $interface interface remote-as internal\""
+}
```
As commented on another commit, I'd like MicroOVN to take responsibility of this configuration.

mkalcok replied:
With the addition of the `asn` option, we now have tests that test both "automatic" and "manual" FRR configuration.

fnordahl commented on tests/test_helper/bats/bgp.bats:
```diff
@@ -0,0 +1,116 @@
+        echo "# ($MICROOVN_BGP_CONTAINER) waiting on established BGP with $container" >&3
+        wait_until "microovn_bgp_established $MICROOVN_BGP_CONTAINER $vrf_device $container"
+    done
+}
```
Should we also check redistribution of routes? At this stage it could be as simple as adding a couple of NAT entries and checking that our bare bones BGP config allows the route to this address show up in the routing table in the remote BGP peer.

mkalcok replied:
This was not yet added.

fnordahl commented on tests/test_helper/setup_teardown/bgp.bash:
```diff
@@ -4,6 +4,8 @@ setup_file() {
     load test_helper/bgp_utils.bash
 
+    # Ensure that required VRF kernel module is available on the host
+    lsmod | awk '{print $1}'| grep -i "\<vrf\>"
```
See above comment on using LXD profile to load the module.

mkalcok replied:
Thank you, yes, that was indeed better approach. This was removed.

fnordahl commented on .github/workflows/tests.yml:
```diff
@@ -130,6 +130,8 @@ jobs:
           sudo lxd init --auto
           snap list
+          sudo apt install -y linux-modules-extra-$(uname -r)
+          sudo modprobe vrf
```
We have a precedent in the code base for installing kernel modules through LXD:
https://github.com/canonical/microovn/blob/31ce8c71de4f3ab6074d71ad74579a2d73da31a2/tests/lxd-instance-config.yaml#L3

Using that would make loading automatic regardless of environment, and presumably it would also provide a useful error message if the module is not available?

mkalcok replied:
Excellent point! thank you. I didn't realize that we can use this mechanism. The `lxc launch` fails with
```
Error: Failed to load kernel module 'vrf': Failed to run: modprobe -b vrf: exit status 1
```
if the module is not available.

mkalcok replied:
I had bit of a dilemma here as I didn't want to enforce `vrf` kernel module for all containers and I didn't want to duplicate the profile either. I ended up using `launch_containers_args` function that allows us to pass `-c linux.kernel_modules=vrf,openvswitch` arg that overrides the profile just for the containers that need the vrf.

My concern with requiring VRF for all containers was that it' only needed in single test suite and reuqires installation of quite heavy package `linux-modules-extra`. Given that we distribute our test runs across many machines, most of them would install this package unnecessarily.

fnordahl reviewed (changes requested):
Martin, first and foremost thank you so much for your work on this feature.

Making all of the in-flight pieces come together like this is non-trivial work, and I really appreciate the care you have taken for everything from UX, adding documentation and end to end tests.

Being able to observe all the pieces together like this is of utmost importance for us to figure out the details of what the final implementation will look like, and you have unlocked our ability to be able to experiment and iterate our way towards a great end to end feature.

With some small adjustments we can probably put this in the hands of interested early adopters too and unlock their thoughts and ideas on how we could make this useful.

I have some comments in-line, please have a look.

Given the size of the PR I have prioritized the areas where I saw the most potential for improvement, so there are other parts where I still may have opinions but we need to prioritize what to address now and what to address as we iterate towards the final product in the coming cycle.

mkalcok commented on snap/snapcraft.yaml:
```diff
@@ -180,6 +205,129 @@ parts:
       - lib/libdqlite*so*
       - lib/libraft*so*
       - lib/*/libuv*so*
+  rtrlib:
+    build-packages:
+      - cmake
+      - make
+      - gcc
+      - libssh-dev
+    stage-packages:
+      - libssh-4
+      - zlib1g
+    prime:
+      - lib/librtr.so*
+      - usr/lib/$SNAPCRAFT_ARCH_TRIPLET/libssh.so*
+    source: https://github.com/rtrlib/rtrlib.git
+    source-type: git
+    source-tag: v0.8.0
+    plugin: cmake
+    cmake-parameters:
+      - -DCMAKE_BUILD_TYPE=Release
+  libyang:
+    build-packages:
+      - cmake
+      - make
+      - gcc
+      - libpcre2-dev
+    stage-packages:
+      - libpcre2-8-0
+    source: https://github.com/CESNET/libyang.git
+    source-type: git
+    source-tag: v2.1.128
+    plugin: cmake
+    cmake-parameters:
+      - -DCMAKE_INSTALL_PREFIX:PATH=/usr
+      - -DENABLE_LYD_PRIV=ON
+      - -DENABLE_CACHE=ON
+      - -DCMAKE_BUILD_TYPE:String="Release"
```
You are correct. The reason is that FRR [ requires for libyang `2.1.128`](https://docs.frrouting.org/projects/dev-guide/en/latest/building-frr-for-ubuntu2204.html#installing-dependencies)

mkalcok commented on snap/hooks/install:
```diff
@@ -18,3 +18,13 @@ supported.
 fi
+
+
+# Create directories required by FRR
```
You are right, thanks, I did not realize that.

mkalcok replied:
As part of this move, I'll change the `cp -r` below to `cp -r --update=none`, to avoid overwriting existing config files on snap refresh

mkalcok replied:
I ended up doing this step both in `install` and in `post-refresh` hook, because it turned out that the `post-refresh` is not executed on fresh install.

mkalcok commented on snap/snapcraft.yaml:
```diff
@@ -263,7 +411,7 @@ parts:
 
       # Check that `golangci-lint` is happy with the code.
-      golangci-lint run --verbose
+       golangci-lint run --verbose
```
nope, it sneaked in when I was "undoing" the linter skip :D. I'll fix it

mkalcok commented on snap/snapcraft.yaml:
```diff
@@ -347,3 +495,9 @@ parts:
       - lib/*/libjq.so*
       - lib/*/libonig.so*
+
+layout:
```
The `/var/tmp/frr` path [is hardcoded](https://github.com/FRRouting/frr/blob/fcc8738bb505b80bd799b4218eea9b7517093294/lib/zlog.c#L1076) so we won't get around that.

We could compile with custom `--sysconfigdir`, but if we ever switch to FRR from package, we'll have one less thing to do if we just leave both of these layouts in place now.

mkalcok commented on microovn/ovn/cmd/commands.go:
```diff
@@ -38,23 +40,69 @@ type OvsdbSpec struct {
+	if dbType == OvsdbTypeNBLocal || dbType == OvsdbTypeNBCluster {
+		var url string
+		if dbType == OvsdbTypeNBLocal {
+			url = fmt.Sprintf("unix:%s", paths.OvnNBDatabaseSock())
```
Yes, the Local/Cluster versions of Northbound differ only in the socket, so I didn't want to copy-paste the entire struct initialization. Golang does not support ternary operators, so the only way that I can think of to "shorten" this would be
```golang
url := fmt.Sprintf("unix:%s", paths.OvnNBDatabaseSock())
if dbType == OvsdbTypeNBCluster {
    	url, err = loadConnectionString(dbType)
    	if err != nil {
    	    return nil, err
    	}
}
```
However with your suggestion above to use command wrappers to talk to clustered OVN NB, this may get removed entirely.

mkalcok replied:
Removed entirely in favor of using `ovn-nbctl` wrapper.

mkalcok commented on microovn/ovn/cmd/commands.go:
```diff
@@ -86,20 +134,42 @@ func NewOvsdbSpec(dbType OvsdbType) (*OvsdbSpec, error) {
+// If SocketURL specified in "db" parameter points to a local unix socket, it does not need to exist before
+// this function is executed, creation of the socket  file will be awaited as well.
 func WaitForDBState(ctx context.Context, _ state.State, db *OvsdbSpec, dbState string, timeout int) error {
+	args := []string{
+		"--timeout",
+		strconv.Itoa(timeout),
+	}
+
+	socketURL := db.SocketURL
+	// Insert certificates arguments if DB connection uses SSL
+	if strings.HasPrefix(db.SocketURL, "ssl") {
+		args = append(args, certArgs()...)
+		// ovsdb-client is not capable to connect to clustered database,
+		// so we select just one of the SSL endpoints
+		socketURL, _, _ = strings.Cut(socketURL, ",")
+	}
+	args = append(args, "wait", socketURL, db.Name, dbState)
+
```
Yes, this function suffers a bit because it's used by many commands in this package (nbctl/sbctl/vsctl/ovsdb-client), so it has to service many scenarios. I'm not happy with the ugly workaround using first IP either, I'll investigate how to improve this.

mkalcok replied:
Now that the `NBCtlCluster` uses `ovn-nbctl` wrapper, this function went back to being used only for local unix sockets.

mkalcok commented on microovn/api/types/services.go:
```diff
@@ -99,10 +102,103 @@ const (
+type ExtraBgpConfig struct {
+	// ExternalConnection is comma separated list of <iface_name>:<ip4_cidr> values. "iface_name"
+	// is a name of the physical interface that provides connectivity to the external network and
+	// "ip4_cidr" is IPv4 address (e.g. 10.0.0.1/24) that should be assigned to a Logical Router
```
I did not know that this was a thing :flushed: thanks.

mkalcok commented on microovn/api/types/services.go:
```diff
@@ -99,10 +102,103 @@ const (
+	ExternalConnection string `json:"ext_iface,omitempty" yaml:"ext_iface,omitempty"`
+	// Vrf is a VRF table ID into which the OVN will leak its routes
+	Vrf string `json:"vrf,omitempty" yaml:"vrf,omitempty"`
+}
```
Do you expect that the final implementation of BGP integration in OVN will require us to configure ASNs at some point? Because currently the ASN is only needed once you are configuring the daemon itself (in `frr`), and that's not something MicroOVN is doing. MicroOVN only sets up the OVN BGP redirection, and the rest of the daemon configuration is up to the user.

mkalcok replied:
`Asn` was added as an optional config parameter. When specified, MicroOVN will perform iBGP unnumbered configuration, as you suggested.

mkalcok commented on microovn/frr/bgp/redirect.go:
```diff
@@ -238,28 +284,157 @@ func redirectBgp(ctx context.Context, s state.State, extConnections []types.BgpE
 			"add-port", intBr, bgpIface,
 			"--",
+			"set", "Port", bgpIface, fmt.Sprintf("external-ids:%s=true", BgpManagedTag),
+			fmt.Sprintf("external-ids:%s=%s", BgpVrfTable, vrfName), fmt.Sprintf("external-ids:%s=%s", BgpIfaceIP, bgpIfaceIP4),
```
Hmm, it may get bit complicated, but it should be possible. 

> We create and maintain this port as a side effect of there being a options:routing-protocol-redirect key present on the LRP whose value is the name of this port.

Value of this option is the LSP name, not the port. However what we could do, during the restart/reboot, we could iterate over all `Port`s to find those tagged with `BgpManagedTag` and for each:
* get associated `Interface`
* get `iface-id` from the `Interface` to identify LSP
* get LSP's peer to identify LRP
* get IP from LRP
* find LRP's LR
* get `requested-tnl-key` from LR

I'm bit on the fence whether it's worth it doing it this way. What do you think?

mkalcok replied:
This remains unchanged. The alternative approach seems bit too complicated to me. Especially given that we don't (currently) support reconfiguration, only enable/disable. So I don't see much opportunity for these values to get out of sync

mkalcok commented on microovn/frr/bgp/redirect.go:
```diff
@@ -0,0 +1,265 @@
+func getOvnIntegrationBridge(ctx context.Context, s state.State) string {
+	brName, err := ovnCmd.VSCtl(ctx, s, "get", "Open_vSwitch", ".", "external-ids:ovn-bridge")
+	if err != nil {
+		return "br-int"
```
Return codes for "no record" is the same as for "can't connect to db", but we could parse the stderr message for `no key "ovn-bridge"` string. what do you think?

mkalcok replied:
This is still unsolved as I'm not sure about the path forward. I think we could go forward even with hardcoded `br-int`, what would be a scenario in which the value changes to non-default? Do CMS mess with the integration bridge name?

mkalcok commented on microovn/client/client.go:
```diff
@@ -157,11 +157,11 @@ func DisableService(ctx context.Context, c *client.Client, serviceName string) (
 // EnableService sends request to disable service with name as as specified in
 // "serviceName" argument.
-func EnableService(ctx context.Context, c *client.Client, serviceName string) (types.WarningSet, types.RegenerateEnvResponse, error) {
+func EnableService(ctx context.Context, c *client.Client, serviceName string, extraConfig *types.ExtraServiceConfig) (types.WarningSet, types.RegenerateEnvResponse, error) {
```
I struggled a bit to draw a line where it would make sense to separate the commit. Although I tried to make the `--config` feature as reusable possible, some of the parts (e.g. [parsing the args into struct](https://github.com/mkalcok/microovn/blob/837db1e1b26fa124b2670a90d584d257d6edb17c/microovn/cmd/microovn/service_control.go#L145-L154)) would look weird or require placeholders if there's no subcommand that's actually using it.

However I can give it another go.

mkalcok replied:
This remains unsolved. Let me know if you think it's a blocker, because atm I don't see a nice way to split this commit.

mkalcok commented on snapcraft/commands/frr-zebra.start:
```diff
@@ -0,0 +1,5 @@
+# Load the environment
+. "${SNAP}/frr.env"
+
+exec zebra $_FRR_COMMON_ARGS --pid_file "$_FRR_RUNDIR/zebra.pid" "$@"
```
fixed

mkalcok commented on snapcraft/commands/vtysh:
```diff
@@ -0,0 +1,5 @@
+# Load the environment
+. "${SNAP}/frr.env"
+
+exec vtysh --vty_socket "$_FRR_RUNDIR" "${@}"
```
fixed

mkalcok commented on snapcraft/frr.env:
```diff
@@ -0,0 +1,10 @@
+                         --socket "$_FRR_RUNDIR/zebra.sock" \
+                         --config_file "$SNAP_COMMON/data/frr/etc/frr.conf" \
+                         --vty_port 0
+                         --limit-fds 100000"
```
fixed

mkalcok commented on snap/snapcraft.yaml:
```diff
@@ -330,13 +330,50 @@ parts:
+        --sysconfdir=/etc \
+        --prefix=/usr
+      make -j$CRAFT_PARALLEL_BUILD_COUNT
+      make install DESTDIR=$CRAFT_PART_INSTALL
```
`ovs` is now regular "part" again, and OVS submodule is used only for building OVN

mkalcok commented on microovn/ovn/cmd/commands.go:
```diff
@@ -38,23 +40,69 @@ type OvsdbSpec struct {
+// loadConnectionString returns string that defines network connection to OVSDB. Information
+// is sourced from the OVN environment file (paths.OvnEnvFile).
+// Example of returned value: "ssl:10.0.0.1:6641,ssl:10.0.0.2:6641,ssl:10.0.0.3:6641"
+func loadConnectionString(dbType OvsdbType) (string, error) {
```
Removed completely in favor of using `ovn-nbctl` wrapper.

mkalcok commented on microovn/ovn/cmd/commands.go:
```diff
@@ -38,23 +40,69 @@ type OvsdbSpec struct {
 	var err error
 
-	if dbType == OvsdbTypeNBLocal {
+	if dbType == OvsdbTypeNBLocal || dbType == OvsdbTypeNBCluster {
```
`OvsdbTypeNBCluster` was removed completely in favor of using `ovn-nbctl` wrapper to contact clustered NB DB

mkalcok commented on microovn/frr/bgp/redirect.go:
```diff
@@ -0,0 +1,265 @@
+//	types.BgpExternalConnection.IPAddress: 192.168.10.1
+//	types.BgpExternalConnection.IPMask:    255.255.255.0
+//
+// Result: "192.168.10.1/24"
```
replaced

mkalcok commented on microovn/frr/bgp/redirect.go:
```diff
@@ -0,0 +1,265 @@
+// Warning: There is no guarantee that the address won't conflict with other MAC addresses
+// present in the network.
+func generateLrpMac(ifaceName string) string {
+	macAddr := "02:"
```
This remains unchanged for now.

fnordahl commented on microovn/frr/bgp/redirect.go:
```diff
@@ -238,28 +284,157 @@ func redirectBgp(ctx context.Context, s state.State, extConnections []types.BgpE
+	}
+
+	// Cleanup ovn-bridge mappings for external networks
+	ovnBridgeMapping, err := ovnCmd.VSCtl(ctx, s, "get", "Open_vSwitch", ".", "external-ids:ovn-bridge-mappings")
```
This call creates an error when it does not exist, which is the case when a user has ran `microovn enable bgp` without options only to then do `microovn disable bgp`.

mkalcok commented:
Thanks for catching these issues. I fixed them and updated tests.
However I found something that I can't explain yet. In tests that enable BGP on multiple chassis (regardless of whether it's with single or with two external neighbours), some of the BGP sessions are stuck in `Open (waiting for connection)` state and never reach `Established`. So far I can't say why that is, as there  doesn't appear to be anything wrong with the setup. I'll dig deeper into it, currently a workaround is to introduce few seconds of sleep between enabling the BGP services on MicroOVN nodes. 
This is of course not a valid solution, so I'll put the PR into `Draft` stage, but I think it's otherwise ready for another round of review.

fnordahl commented:
> Thanks for catching these issues. I fixed them and updated tests. However I found something that I can't explain yet. In tests that enable BGP on multiple chassis (regardless of whether it's with single or with two external neighbours), some of the BGP sessions are stuck in `Waiting for peer OPEN (n/a)` state and never reach `Established`. So far I can't say why that is, as there doesn't appear to be anything wrong with the setup. I'll dig deeper into it, currently a workaround is to introduce few seconds of sleep between enabling the BGP services on MicroOVN nodes. This is of course not a valid solution, so I'll put the PR into `Draft` stage, but I think it's otherwise ready for another round of review.

I have not looked yet, but from your description of BGP sessions failing to form, my hunch would be https://github.com/FRRouting/frr/issues/8668#issuecomment-910770769.

Essentially there can only be one unnumbered BGP peer per interface, and the test uses a single L2 broadcast domain for all the containers, so I think this is what we see here.

I guess we could have the test create more LXD networks, one per host-ToR link, to simulate the point to point nature of a physical cable?

mkalcok commented:
Thanks for the suggestion, but unfortunately, I don't think that this is the problem. Tests already create individual networks for each ToR <-> Chassis connection. (Given that this is indeed pretty complex setup, I included more details about it on top of the `setup_file` for `bgp` tests)

fnordahl commented:
> Thanks for the suggestion, but unfortunately, I don't think that this is the problem. Tests already create individual networks for each ToR <-> Chassis connection. (Given that this is indeed pretty complex setup, I included more details about it on top of the `setup_file` for `bgp` tests)

Indeed you do, that's awesome, and sorry for missing that. I'll have a peek and see if I spot anything.

mkalcok commented:
I changed my mind on the topic of merging this PR. Given that this is going to be an experimental feature branch (not main or stable), I think it's OK to merge this even with the unresolved issue mentioned in the [comment above](https://github.com/canonical/microovn/pull/182#issuecomment-2389716245). The issue seems to manifest only when BGP is enabled in quick succession on multiple chassis. Any deployments configured by hand (or with slight backoff) should work fine.

If this ends up merged, I'll open a LP ticket regarding this issue and we can look at it later. (I have a slight suspicion that it may be even an upstream OVN issue)

fnordahl reviewed (approved):
There are a few unresolved conversations on this PR, being a feature branch I think that is fine as we don't know which parts of this will survive into the final version yet. Let's revisit once we integrate back to main.

LGTM, thank you for your work on this!
